zap/src/main/java/org/zaproxy/zap/session/ScriptBasedSessionManagementMethodType.java

```diff
@@ -0,0 +1,739 @@
+import java.awt.Component;
+import java.awt.GridBagLayout;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.swing.DefaultComboBoxModel;
+import javax.swing.ImageIcon;
+import javax.swing.JButton;
+import javax.swing.JComboBox;
+import javax.swing.JLabel;
+import javax.swing.JList;
+import javax.swing.JOptionPane;
+import javax.swing.JPanel;
+import javax.swing.border.Border;
+import javax.swing.border.EmptyBorder;
+import javax.swing.plaf.basic.BasicComboBoxRenderer;
+import net.sf.json.JSON;
+import net.sf.json.JSONObject;
+import org.apache.commons.configuration.Configuration;
+import org.apache.commons.configuration.ConfigurationException;
+import org.apache.commons.httpclient.HttpState;
+import org.apache.commons.lang.StringEscapeUtils;
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.db.DatabaseException;
+import org.parosproxy.paros.db.RecordContext;
+import org.parosproxy.paros.extension.ExtensionHook;
+import org.parosproxy.paros.model.Session;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.view.View;
+import org.zaproxy.zap.extension.api.ApiDynamicActionImplementor;
+import org.zaproxy.zap.extension.api.ApiException;
+import org.zaproxy.zap.extension.api.ApiResponse;
+import org.zaproxy.zap.extension.api.ApiResponseSet;
+import org.zaproxy.zap.extension.script.ExtensionScript;
+import org.zaproxy.zap.extension.script.ScriptType;
+import org.zaproxy.zap.extension.script.ScriptWrapper;
+import org.zaproxy.zap.extension.sessions.ExtensionSessionManagement;
+import org.zaproxy.zap.extension.sessions.SessionManagementAPI;
+import org.zaproxy.zap.model.Context;
+import org.zaproxy.zap.utils.ApiUtils;
+import org.zaproxy.zap.utils.EncodingUtils;
+import org.zaproxy.zap.view.DynamicFieldsPanel;
+import org.zaproxy.zap.view.LayoutHelper;
+
+/**
+ * The type corresponding to a {@link SessionManagementMethod} for web applications that uses
+ * Session Management Scripts, which can hopefully model any custom behaviour required
+ *
+ * @since TODO add version
+ */
+public class ScriptBasedSessionManagementMethodType extends SessionManagementMethodType {
+
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT =
+            ExtensionSessionManagement.CONTEXT_CONFIG_SESSION + ".script";
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT_NAME =
+            CONTEXT_CONFIG_SESSION_MGMT_SCRIPT + ".name";
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT_PARAMS =
+            CONTEXT_CONFIG_SESSION_MGMT_SCRIPT + ".params";
+
+    private static final int METHOD_IDENTIFIER = 2;
+
+    private static final Logger lOG =
```

Comment:
Oops 2/3

Reply:
Thats not bad is it?? (obligatory MeatLoaf reference :D)
Now fixed ;)